src/signet.cpp

```diff
@@ -0,0 +1,118 @@
+// Copyright (c) 2019-2020 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <signet.h>
+
+#include <consensus/merkle.h>
+#include <consensus/params.h>
+#include <consensus/validation.h>
+#include <hash.h>
+#include <primitives/block.h>
+#include <primitives/transaction.h>
+#include <script/interpreter.h>
+#include <script/standard.h>
+#include <util/system.h>
+
+// Signet block solution checker
+bool CheckBlockSolution(const CBlock& block, const Consensus::Params& consensusParams)
+{
+    std::vector<uint8_t> signet_data;
+    if (!GetWitnessCommitmentSection(block, SIGNET_HEADER, signet_data)) {
+        return error("CheckBlockSolution: Errors in block (block solution missing)");
+    }
+    SimpleSignatureChecker bsc(GetSignetHash(block));
+    CScript challenge(consensusParams.signet_challenge.begin(), consensusParams.signet_challenge.end());
+    CScript solution = CScript(signet_data.begin(), signet_data.end());
+
+    if (!VerifyScript(solution, challenge, nullptr, MANDATORY_SCRIPT_VERIFY_FLAGS, bsc)) {
+        return error("CheckBlockSolution: Errors in block (block solution invalid)");
+    }
+    return true;
+}
+
+uint256 BlockSignetMerkleRoot(const CBlock& block, bool* mutated = nullptr)
+{
+    std::vector<uint256> leaves;
+    leaves.resize(block.vtx.size());
+    {
+        // find and delete signet signature
+        CMutableTransaction mtx(*block.vtx.at(0));
+        (void)SetWitnessCommitmentSection(mtx, SIGNET_HEADER, std::vector<uint8_t>{});
+        leaves[0] = mtx.GetHash();
+    }
+    for (size_t s = 1; s < block.vtx.size(); ++s) {
+        leaves[s] = block.vtx[s]->GetHash();
+    }
+    return ComputeMerkleRoot(std::move(leaves), mutated);
+}
+
+uint256 GetSignetHash(const CBlock& block)
+{
+    if (block.vtx.size() == 0) return block.GetHash();
+    return (CHashWriter(SER_DISK, PROTOCOL_VERSION) << block.nVersion << block.hashPrevBlock << BlockSignetMerkleRoot(block) << block.nTime << block.nBits).GetHash();
+}
+
+bool GetWitnessCommitmentSection(const CBlock& block, const uint8_t header[4], std::vector<uint8_t>& result)
+{
+    int cidx = GetWitnessCommitmentIndex(block);
+    if (cidx == NO_WITNESS_COMMITMENT) return false;
```

Comment:
This may be more of a note on BIP325 in general, but on mainnet currently it is not invalid to produce a block with no transactions and therefore no witness commitment in the coinbase tx.

I found a recent example: https://blockstream.info/block/0000000000000000000e77c9c81498be9a4aa226d4be35a19a0fb11d99331d6a

Why does the spec depend on a witness commitment instead of new `OP_RETURN` for this purpose?

Reply:
That's a good point, but then again, it IS allowed to put a witness commitment in a block with no transactions (as evidenced by most of the signet blocks right now), so it doesn't feel like a huge deal in the end. After all, the signet blocks will "suffer" from the extra overhead with empty blocks, but that seems like an acceptable trade-off.

The question is if it's simpler to do an OP_RETURN instead of a witness commitment extension. @sipa or someone, would love your input on this.